File: implementations/csharp/TestRunner.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using Hl7.Fhir.ElementModel;
using Hl7.Fhir.FhirPath;
using Hl7.Fhir.Model;
using Hl7.Fhir.Serialization;
using Hl7.FhirPath;

namespace FhirPathComparison
{
    /// <summary>
    /// C# FHIRPath Test Runner
    ///
    /// This class runs FHIRPath tests using the Hl7.FhirPath library
    /// and outputs results in a standardized format for comparison.
    /// </summary>
    public class TestRunner
    {
        private readonly string _specsDir;
        private readonly string _resultsDir;
        private readonly FhirJsonParser _jsonParser;
        private readonly FhirPathCompiler _fhirPathCompiler;

        public TestRunner()
        {
            // Initialize paths
            var currentDir = Directory.GetCurrentDirectory();
            _specsDir = Path.Combine(currentDir, "..", "..", "specs");
            _resultsDir = Path.Combine(currentDir, "..", "..", "results");

            // Ensure results directory exists
            Directory.CreateDirectory(_resultsDir);

            // Initialize FHIR parsers and compiler
            _jsonParser = new FhirJsonParser();
            _fhirPathCompiler = new FhirPathCompiler();
        }

        /// <summary>
        /// Load FHIRPath test cases from new JSON format.
        /// </summary>
        private List<JsonObject> LoadTestSuites()
        {
            var tests = new List<JsonObject>();
            var testsDir = Path.Combine(_specsDir, "fhirpath", "tests");

            if (!Directory.Exists(testsDir))
            {
                Console.WriteLine($"❌ Tests directory not found: {testsDir}");
                return tests;
            }

            try
            {
                var jsonFiles = Directory.GetFiles(testsDir, "*.json");
                
                foreach (var jsonFile in jsonFiles)
                {
                    var testSuite = JsonNode.Parse(File.ReadAllText(jsonFile));
                    var suiteName = testSuite["name"]?.ToString() ?? Path.GetFileNameWithoutExtension(jsonFile);
                    
                    var testCases = testSuite["tests"]?.AsArray();
                    if (testCases == null) continue;
                    
                    foreach (var testCase in testCases)
                    {
                        // Skip disabled tests
                        if (testCase["disable"]?.GetValue<bool>() == true)
                            continue;
                        
                        var inputFile = testCase["inputfile"]?.ToString() ?? "patient-example.json";
                        
                        var test = new JsonObject
                        {
                            ["name"] = testCase["name"]?.ToString(),
                            ["description"] = testCase["description"]?.ToString() ?? testCase["name"]?.ToString(),
                            ["inputFile"] = inputFile,
                            ["expression"] = testCase["expression"]?.ToString(),
                            ["group"] = suiteName,
                            ["invalid"] = testCase["error"] != null,
                            ["expectedOutput"] = testCase["expected"]
                        };

                        tests.Add(test);
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"❌ Error loading test suites: {e.Message}");
            }

            return tests;
        }

        /// <summary>
        /// Load test data from JSON file.
        /// </summary>
        private Resource LoadTestData(string filename)
        {
            var filePath = Path.Combine(_specsDir, "fhirpath", "tests", "input", filename);

            if (!File.Exists(filePath))
            {
                Console.WriteLine($"⚠️  Test data file not found: {filename}");
                return null;
            }

            try
            {
                var jsonContent = File.ReadAllText(filePath);
                var parser = _jsonParser;
                var resource = parser.Parse<Resource>(jsonContent);
                return resource;
            }
            catch (Exception e)
            {
                Console.WriteLine($"⚠️  Error loading test data {filename}: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// Run a single test case and return results.
        /// </summary>
        private JsonObject RunSingleTest(JsonNode testCase, Resource testData)
        {
            var result = new JsonObject();
            var stopwatch = Stopwatch.StartNew();

            try
            {
                var expression = testCase["expression"]?.ToString();

                // Execute FHIRPath expression
                var compiled = _fhirPathCompiler.Compile(expression);
                var evaluationResult = compiled(testData.ToTypedElement(), EvaluationContext.CreateDefault());

                stopwatch.Stop();
                var executionTimeMs = stopwatch.Elapsed.TotalMilliseconds;

                result["name"] = testCase["name"]?.ToString();
                result["description"] = testCase["description"]?.ToString();
                result["expression"] = expression;
                result["status"] = "passed"; // Simplified - would need proper result comparison
                result["execution_time_ms"] = executionTimeMs;

                // Add expected and actual results
                if (testCase["expectedOutput"] != null)
                {
                    result["expected"] = JsonNode.Parse(testCase["expectedOutput"].ToJsonString());
                }
                else
                {
                    result["expected"] = new JsonArray();
                }

                var actualArray = new JsonArray();
                foreach (var item in evaluationResult)
                {
                    if (item.Value != null)
                    {
                        actualArray.Add(JsonValue.Create(item.Value.ToString()));
                    }
                    else
                    {
                        actualArray.Add(JsonValue.Create(item.ToString()));
                    }
                }
                result["actual"] = actualArray;
            }
            catch (Exception error)
            {
                stopwatch.Stop();
                var executionTimeMs = stopwatch.Elapsed.TotalMilliseconds;

                result["name"] = testCase["name"]?.ToString();
                result["description"] = testCase["description"]?.ToString();
                result["expression"] = testCase["expression"]?.ToString();
                result["status"] = "error";
                result["execution_time_ms"] = executionTimeMs;

                if (testCase["expectedOutput"] != null)
                {
                    result["expected"] = JsonNode.Parse(testCase["expectedOutput"].ToJsonString());
                }
                else
                {
                    result["expected"] = new JsonArray();
                }
                result["actual"] = null;
                result["error"] = error.Message;
            }

            return result;
        }

        /// <summary>
        /// Run all tests and return results.
        /// </summary>
        public JsonObject RunTests()
        {
            Console.WriteLine("🧪 Running C# FHIRPath tests...");

            var results = new JsonObject
            {
                ["language"] = "csharp",
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                ["tests"] = new JsonArray(),
                ["summary"] = new JsonObject
                {
                    ["total"] = 0,
                    ["passed"] = 0,
                    ["failed"] = 0,
                    ["errors"] = 0
                }
            };

            var testsArray = results["tests"].AsArray();
            var summary = results["summary"].AsObject();

            // Load and run test suites
            Console.WriteLine("📋 Loading FHIRPath test suites...");
            var testCases = LoadTestSuites();
            Console.WriteLine($"📊 Found {testCases.Count} test cases");

            foreach (var testCase in testCases)
            {
                var inputFile = testCase["inputFile"].ToString();
                
                // Load test data on demand
                var testData = !string.IsNullOrEmpty(inputFile) ? LoadTestData(inputFile) : null;
                
                if (testData == null)
                {
                    Console.WriteLine($"⚠️  Skipping test {testCase["name"]} - test data not available: {inputFile}");
                    continue;
                }

                // Skip tests marked as invalid for now (these test error conditions)
                if (testCase["invalid"].GetValue<bool>())
                {
                    Console.WriteLine($"⏭️  Skipping invalid test {testCase["name"]} (tests error conditions)");
                    continue;
                }

                var testResult = RunSingleTest(testCase, testData);
                testsArray.Add(testResult);

                var total = summary["total"].GetValue<int>() + 1;
                summary["total"] = total;

                var status = testResult["status"].ToString();
                if (status == "passed")
                {
                    summary["passed"] = summary["passed"].GetValue<int>() + 1;
                }
                else if (status == "error")
                {
                    summary["errors"] = summary["errors"].GetValue<int>() + 1;
                }
                else
                {
                    summary["failed"] = summary["failed"].GetValue<int>() + 1;
                }

                var statusIcon = status == "passed" ? "✅" :
                                status == "error" ? "💥" : "❌";
                var executionTime = testResult["execution_time_ms"].GetValue<double>();
                var groupName = testCase["group"].ToString();
                Console.WriteLine($"  {statusIcon} {testResult["name"]} ({executionTime:F2}ms) [{groupName}]");
            }

            // Save results
            var resultsFile = Path.Combine(_resultsDir, "csharp_test_results.json");
            var options = new JsonSerializerOptions { WriteIndented = true };
            File.WriteAllText(resultsFile, results.ToJsonString(options));

            Console.WriteLine($"📊 Results saved to: {resultsFile}");
            Console.WriteLine($"📈 Summary: {summary["passed"]}/{summary["total"]} tests passed");

            return results;
        }

        /// <summary>
        /// Run benchmarks and return results.
        /// </summary>
        public JsonObject RunBenchmarks()
        {
            Console.WriteLine("⚡ Running C# FHIRPath benchmarks...");

            var results = new JsonObject
            {
                ["language"] = "csharp",
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                ["benchmarks"] = new JsonArray(),
                ["system_info"] = new JsonObject
                {
                    ["platform"] = Environment.OSVersion.ToString(),
                    ["dotnet_version"] = Environment.Version.ToString(),
                    ["fhirpath_version"] = "Hl7.FhirPath 5.5.0" // Would need to get actual version
                }
            };

            var benchmarksArray = results["benchmarks"].AsArray();

            // Create simple benchmarks from test cases
            var testCases = LoadTestSuites();
            var benchmarkTests = new List<JsonObject>();
            
            // Use a subset of test cases for benchmarking
            for (int i = 0; i < Math.Min(10, testCases.Count); i++)
            {
                var testCase = testCases[i];
                var benchmarkTest = new JsonObject
                {
                    ["name"] = $"benchmark_{testCase["name"]}",
                    ["description"] = $"Benchmark for {testCase["name"]}",
                    ["expression"] = testCase["expression"],
                    ["inputFile"] = testCase["inputFile"],
                    ["iterations"] = 100
                };
                benchmarkTests.Add(benchmarkTest);
            }

            // Run benchmarks
            foreach (var benchmark in benchmarkTests)
            {
                var inputFile = benchmark["inputFile"]?.ToString() ?? "patient-example.json";
                
                // Load test data on demand
                var testData = !string.IsNullOrEmpty(inputFile) ? LoadTestData(inputFile) : null;
                
                if (testData == null)
                {
                    Console.WriteLine($"⚠️  Skipping benchmark {benchmark["name"]} - test data not available");
                    continue;
                }
                Console.WriteLine($"  🏃 Running {benchmark["name"]}...");

                var expression = benchmark["expression"].ToString();
                var iterations = benchmark["iterations"]?.GetValue<int>() ?? 1000;
                var times = new List<double>();

                try
                {
                    var compiled = _fhirPathCompiler.Compile(expression);
                    var typedElement = testData.ToTypedElement();
                    var context = EvaluationContext.CreateDefault();

                    // Warm up
                    for (int i = 0; i < 10; i++)
                    {
                        try
                        {
                            compiled(typedElement, context).ToList();
                        }
                        catch
                        {
                            // Ignore warm-up errors
                        }
                    }

                    // Actual benchmark
                    for (int i = 0; i < iterations; i++)
                    {
                        var stopwatch = Stopwatch.StartNew();
                        try
                        {
                            compiled(typedElement, context).ToList();
                        }
                        catch
                        {
                            // Continue timing even if expression fails
                        }
                        stopwatch.Stop();
                        times.Add(stopwatch.Elapsed.TotalMilliseconds);
                    }

                    if (times.Any())
                    {
                        var avgTime = times.Average();
                        var minTime = times.Min();
                        var maxTime = times.Max();
                        var opsPerSecond = avgTime > 0 ? 1000.0 / avgTime : 0.0;

                        var benchmarkResult = new JsonObject
                        {
                            ["name"] = benchmark["name"].ToString(),
                            ["description"] = benchmark["description"].ToString(),
                            ["expression"] = expression,
                            ["iterations"] = iterations,
                            ["avg_time_ms"] = avgTime,
                            ["min_time_ms"] = minTime,
                            ["max_time_ms"] = maxTime,
                            ["ops_per_second"] = opsPerSecond
                        };

                        benchmarksArray.Add(benchmarkResult);
                        Console.WriteLine($"    ⏱️  {avgTime:F2}ms avg ({opsPerSecond:F1} ops/sec)");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"    ⚠️  Skipping {benchmark["name"]} - expression failed: {ex.Message}");
                    continue;
                }
            }

            // Save results
            var resultsFile = Path.Combine(_resultsDir, "csharp_benchmark_results.json");
            var options = new JsonSerializerOptions { WriteIndented = true };
            File.WriteAllText(resultsFile, results.ToJsonString(options));

            Console.WriteLine($"📊 Benchmark results saved to: {resultsFile}");

            return results;
        }

        public static void Main(string[] args)
        {
            try
            {
                var runner = new TestRunner();
                var command = args.Length > 0 ? args[0] : "both";

                if (command == "test" || command == "both")
                {
                    runner.RunTests();
                }

                if (command == "benchmark" || command == "both")
                {
                    runner.RunBenchmarks();
                }

                Console.WriteLine("✅ C# test runner completed");
            }
            catch (Exception error)
            {
                Console.WriteLine($"❌ Error running tests: {error.Message}");
                Console.WriteLine(error.StackTrace);
                Environment.Exit(1);
            }
        }
    }
}

File: implementations/clojure/src/test_runner.clj
(ns test-runner
  "Clojure FHIRPath Test Runner

  This script runs FHIRPath tests using the fhirpath.clj library
  and outputs results in a standardized format for comparison."
  (:require [clojure.data.json :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [fhirpath.core :as fp])
  (:import [java.time Instant]
           [java.io File]))

(defn current-timestamp []
  (.toString (Instant/now)))

(defn load-test-suites []
  "Load FHIRPath test cases from new JSON format"
  (let [tests-dir (io/file "../../specs/fhirpath/tests")
        tests (atom [])]
    (when (.exists tests-dir)
      (doseq [json-file (.listFiles tests-dir)
              :when (and (.isFile json-file)
                        (.endsWith (.getName json-file) ".json"))]
        (try
          (let [test-suite (json/read-str (slurp json-file) :key-fn keyword)
                suite-name (or (:name test-suite) (str/replace (.getName json-file) ".json" ""))
                test-cases (:tests test-suite [])]
            (doseq [test-case test-cases
                    :when (not (:disable test-case))]
              (let [input-file (or (:inputfile test-case) "patient-example.json")]
                (swap! tests conj {
                  :name (:name test-case)
                  :description (or (:description test-case) (:name test-case))
                  :inputFile input-file
                  :expression (:expression test-case)
                  :expectedOutput (or (:expected test-case) [])
                  :predicate false
                  :mode nil
                  :invalid (some? (:error test-case))
                  :group suite-name
                  :tags (or (:tags test-case) [])
                }))))
          (catch Exception e
            (println (str "⚠️  Failed to load test file " (.getName json-file) ": " (.getMessage e)))))))
    @tests))

(defn load-test-data [filename]
  "Load test data from JSON file"
  (let [file-path (io/file (str "../../specs/fhirpath/tests/input/" filename))]
    (when (.exists file-path)
      (try
        (let [json-content (slurp file-path)
              parsed-json (json/read-str json-content :key-fn keyword)]
          parsed-json)
        (catch Exception e
          (println (str "⚠️  Error loading test data " filename ": " (.getMessage e)))
          nil)))))

(defn safe-serialize [obj]
  "Safely serialize objects to JSON-compatible format"
  (try
    (cond
      (nil? obj) nil
      (string? obj) obj
      (number? obj) obj
      (boolean? obj) obj
      (keyword? obj) (name obj)
      (sequential? obj) (mapv safe-serialize obj)
      (map? obj) (into {} (map (fn [[k v]] [(safe-serialize k) (safe-serialize v)]) obj))
      :else (str obj))
    (catch Exception e
      (str "<non-serializable: " (type obj) ">"))))

(defn run-single-test [test test-data]
  "Run a single FHIRPath test case"
  (let [start-time (System/nanoTime)]
    (try
      (let [result (fp/fp (:expression test) test-data)
            end-time (System/nanoTime)
            execution-time-ms (/ (- end-time start-time) 1000000.0)
            serialized-result (safe-serialize result)]

        {:name (:name test)
         :expression (:expression test)
         :result serialized-result
         :expected (:expectedOutput test)
         :success true
         :executionTimeMs execution-time-ms
         :error nil})

      (catch Exception e
        (let [end-time (System/nanoTime)
              execution-time-ms (/ (- end-time start-time) 1000000.0)]
          {:name (:name test)
           :expression (:expression test)
           :result nil
           :expected (:expectedOutput test)
           :success false
           :executionTimeMs execution-time-ms
           :error (.getMessage e)})))))

(defn run-tests []
  "Run all FHIRPath tests and return results"
  (println "🧪 Running Clojure FHIRPath tests...")

  (let [tests (load-test-suites)
        start-time (current-timestamp)]

    (println (str "📋 Loaded " (count tests) " test cases"))

    (let [results (for [test tests
                       :let [input-file (:inputFile test)
                             test-data (cond
                                       (nil? input-file) {} ; No input file needed
                                       :else (load-test-data input-file))]
                       :when (or (nil? (:inputFile test)) test-data)] ; Skip only if input file specified but failed to load
                   (do
                     (print (str "  Running: " (:name test) "... "))
                     (flush)
                     (let [result (run-single-test test test-data)]
                       (println (if (:success result) "✅" "❌"))
                       result)))

          end-time (current-timestamp)
          successful-tests (count (filter :success results))
          total-tests (count results)
          success-rate (if (> total-tests 0) (double (/ successful-tests total-tests)) 0.0)]

      (println (str "\n📊 Results: " successful-tests "/" total-tests
                   " tests passed (" (format "%.1f" (* success-rate 100)) "%)"))

      {:language "clojure"
       :timestamp end-time
       :tests results
       :summary {:total total-tests
                 :passed successful-tests
                 :failed (- total-tests successful-tests)
                 :errors 0}})))

(defn save-results [results]
  "Save test results to JSON file"
  (let [filename "../../results/clojure_test_results.json"]
    (try
      (spit filename (json/write-str results :indent true))
      (println (str "💾 Results saved to: " filename))
      (catch Exception e
        (println (str "❌ Error saving results: " (.getMessage e)))))))

(defn run-benchmarks []
  "Run benchmarks and return results"
  (println "⚡ Running Clojure FHIRPath benchmarks...")

  (let [test-cases (load-test-suites)
        benchmark-tests (take 10 test-cases)] ; Use first 10 test cases for benchmarking

    (if (empty? benchmark-tests)
      (do
        (println "⚠️  No benchmark tests found in configuration")
        {:language "clojure"
         :timestamp (.toString (Instant/now))
         :benchmarks []
         :system_info {:platform (System/getProperty "os.name")
                       :java_version (System/getProperty "java.version")
                       :clojure_version (clojure-version)
                       :fhirpath_version "fhirpath.clj"}})

      (let [results {:language "clojure"
                     :timestamp (.toString (Instant/now))
                     :benchmarks []
                     :system_info {:platform (System/getProperty "os.name")
                                   :java_version (System/getProperty "java.version")
                                   :clojure_version (clojure-version)
                                   :fhirpath_version "fhirpath.clj"}}

            test-data-cache {}] ; We'll load test data on demand

        (println (str "📋 Running " (count benchmark-tests) " benchmark tests"))

        (let [benchmark-results
              (for [benchmark benchmark-tests
                    :let [input-file (get benchmark :inputFile "patient-example.json")
                          test-data (load-test-data input-file)]]
                (if (nil? test-data)
                  (do
                    (println (str "⚠️  Skipping benchmark " (:name benchmark) " - test data not available"))
                    nil)
                  (do
                    (println (str "  🏃 Running " (:name benchmark) "..."))
                    (let [iterations (get benchmark :iterations 1000)
                          expression (:expression benchmark)

                          ;; Warm up
                          _ (dotimes [_ 10]
                              (try
                                (fp/fp expression test-data)
                                (catch Exception e nil)))

                          ;; Actual benchmark
                          times (for [_ (range iterations)]
                                  (let [start-time (System/nanoTime)]
                                    (try
                                      (fp/fp expression test-data)
                                      (catch Exception e nil))
                                    (let [end-time (System/nanoTime)]
                                      (/ (- end-time start-time) 1000000.0))))

                          valid-times (filter pos? times)
                          avg-time (if (seq valid-times) (/ (reduce + valid-times) (count valid-times)) 0.0)
                          min-time (if (seq valid-times) (apply min valid-times) 0.0)
                          max-time (if (seq valid-times) (apply max valid-times) 0.0)
                          ops-per-second (if (> avg-time 0) (/ 1000.0 avg-time) 0.0)]

                      (println (str "    ⏱️  " (format "%.2f" avg-time) "ms avg ("
                                   (format "%.1f" ops-per-second) " ops/sec)"))

                      {:name (:name benchmark)
                       :description (:description benchmark)
                       :expression expression
                       :iterations iterations
                       :avg_time_ms avg-time
                       :min_time_ms min-time
                       :max_time_ms max-time
                       :ops_per_second ops-per-second}))))

              valid-results (filter some? benchmark-results)
              final-results (assoc results :benchmarks valid-results)]

          ;; Save benchmark results
          (let [filename "../../results/clojure_benchmark_results.json"]
            (try
              (spit filename (json/write-str final-results :indent true))
              (println (str "📊 Benchmark results saved to: " filename))
              (catch Exception e
                (println (str "❌ Error saving benchmark results: " (.getMessage e))))))

          final-results)))))

(defn -main [& args]
  "Main entry point for the test runner"
  (println "🚀 Starting Clojure FHIRPath Test Runner")
  (println "=====================================")

  (let [command (if (seq args) (first args) "both")]

    (try
      (when (or (= command "test") (= command "both"))
        (let [results (run-tests)]
          (save-results results)))

      (when (or (= command "benchmark") (= command "both"))
        (run-benchmarks))

      (println "\n✅ Clojure test runner completed successfully")

      (catch Exception e
        (println (str "\n❌ Test runner failed: " (.getMessage e)))
        (System/exit 1)))))

;; Allow running as script
(when (= *file* (first *command-line-args*))
  (-main))
